File: SocketStorage/Message.cpp
#include "pch.h"
#include "Message.h"

Message::Message(int mTo, int mFrom, int mType, const string& data)
{
	this->data = data;
	header = { mTo, mFrom, mType, (int)data.length() };
}

MessageHeader Message::getHeader()
{
	return header;
}

string Message::getData()
{
	return data;
}

void Message::send(CSocket& s)
{
	s.Send(&header, sizeof(MessageHeader));
	if (header.mSize)
	{
		s.Send(data.c_str(), (int)header.mSize);//+1 ?
	}
}

int Message::receive(CSocket& s) {
	if (!s.Receive(&header, sizeof(MessageHeader)))
	{
		return MT_NODATA;
	}
	if (header.mSize)
	{
		char* v = new char[header.mSize + 1];

		s.Receive(v, header.mSize + 1);
		v[header.mSize] = '\0';
		data = v;

		delete[] v;
	}
	return header.mType;
}

void Message::send(CSocket& s, int mTo, int mFrom, int mType, const string& data) {
	Message m(mTo, mFrom, mType, data);
	m.send(s);
}

// Функция для отправки ответов из хранилища
Message Message::sendArc(unsigned int mType, const string& data) {
	CSocket s;
	s.Create();

	if (!s.Connect(_T("127.0.0.1"), 12345))
	{
		DWORD dwError = GetLastError();
		throw runtime_error("Connection error");
	}

	Message m(MR_USER, MR_STORAGE, mType, data);
	m.send(s);

	return m;
}

File: SocketClientSharp/Message.cs
using System;
using System.Text;
using System.Net.Sockets;

namespace SocketClientSharp
{
    enum MessageRecipients
    {
        MR_STORAGE = -2,
        MR_ALL = -1,
        MR_USER = 0
    };

    enum MessageTypes
    {
        MT_INIT,
        MT_EXIT,
        MT_GETDATA,
        MT_DATA,
        MT_NODATA,
        MT_CONFIRM,
        MT_GETMESSAGES
    };

    struct MessageHeader
    {
        public int mTo;
        public int mFrom;
        public int mType;
        public int mSize;
    };

    class Message
    {
        public string Data { get; set; } = "";

        public MessageHeader header = new MessageHeader { mTo = 0, mFrom = 0, mType = 0, mSize = 0 };

        public Message() {}

        public Message(int to, int from, MessageTypes type = MessageTypes.MT_DATA, string data = "")
        {
            header = new MessageHeader { mTo = to, mFrom = from, mType = (int)type, mSize = data.Length };
            Data = data;
        }

        public MessageHeader getHeader()
        {
            return header;
        }

        public void Send(Socket s)
        {
            s.Send(BitConverter.GetBytes(header.mTo), sizeof(int), SocketFlags.None);
            s.Send(BitConverter.GetBytes(header.mFrom), sizeof(int), SocketFlags.None);
            s.Send(BitConverter.GetBytes((int)header.mType), sizeof(int), SocketFlags.None);
            s.Send(BitConverter.GetBytes(header.mSize), sizeof(int), SocketFlags.None);

            if (header.mSize != 0)
            {
                s.Send(Encoding.UTF8.GetBytes(Data), header.mSize, SocketFlags.None);
            }

        }

        public MessageTypes Receive(Socket s)
        {
            byte[] buffer = new byte[4];
            buffer = new byte[4];
            s.Receive(buffer, sizeof(int), SocketFlags.None);
            header.mTo = BitConverter.ToInt32(buffer, 0);

            buffer = new byte[4];
            s.Receive(buffer, sizeof(int), SocketFlags.None);
            header.mFrom = BitConverter.ToInt32(buffer, 0);

            buffer = new byte[4];
            s.Receive(buffer, sizeof(int), SocketFlags.None);
            header.mType = (int)(MessageTypes)BitConverter.ToInt32(buffer, 0);

            buffer = new byte[4];
            s.Receive(buffer, sizeof(int), SocketFlags.None);
            header.mSize = BitConverter.ToInt32(buffer, 0);

            if (header.mSize != 0)
            {
                buffer = new byte[header.mSize];
                s.Receive(buffer, header.mSize, SocketFlags.None);
                Data = Encoding.UTF8.GetString(buffer, 0, header.mSize);
            }

            return (MessageTypes)(int)header.mType;
        }
    }
}
